datafusion/physical-plan/src/joins/sort_merge_join.rs

```diff
@@ -784,6 +790,29 @@ fn get_corrected_filter_mask(
             corrected_mask.extend(vec![Some(false); null_matched]);
             Some(corrected_mask.finish())
         }
+        JoinType::LeftMark => {
+            for i in 0..row_indices_length {
+                let last_index =
+                    last_index_for_row(i, row_indices, batch_ids, row_indices_length);
+                if filter_mask.value(i) && !seen_true {
```

Comment:
This seems like it could be simplified a bit but I see it follows the same structure in existing code

Reply:
Yeah, I just followed the existing code. So it probably better to simplify it as a later step and maybe do the same to the others.

Reply:
There is ongoing work with SortMergeJoin. 
For LeftMark is it the same filtering rules as for LeftOuter join? 

We need to cover it in fuzz tests

Reply:
> For LeftMark is it the same filtering rules as for LeftOuter join?

It not the same as LeftMark will only output a single row per row in left even if there are multiple passing the filter.

> We need to cover it in fuzz tests

This PR contains fuzz test covering this code.

Reply:
My bad, I was talking on LeftSemi. How LeftMark differs from LeftSemi?

Reply:
It will also produced output for unmatched rows, with the mark column being false.

Reply:
yes, 100% documenting the semantics of Left Mark Join will be very important

Reply:
The paper
http://btw2017.informatik.uni-stuttgart.de/slidesandpapers/F1-10-37/paper_web.pdf
is quite helpful to understand it.

The issue explains as well a for queries like `exists` a join + distinct will be needed, this will surely be less efficient than the mark join.

Reply:
Please correct me if I'm wrong supposedly there are 2 options how to achieve LeftMark

EXISTS
```
SELECT 
    a.id,
    a.name,
    CASE 
        WHEN EXISTS (SELECT 1 FROM table_B AS b WHERE a.id = b.table_a_id) THEN 1
        ELSE 0
    END AS has_match
FROM 
    table_A AS a;
```

LeftOuter
```
SELECT 
    a.id,
    a.name,
    CASE 
        WHEN b.table_a_id IS NOT NULL THEN 1
        ELSE 0
    END AS has_match
FROM 
    table_A AS a
LEFT JOIN 
    table_B AS b
ON 
    a.id = b.table_a_id;
```

The EXISTS approach looks overheated but LeftOuter interpretation  looks for me as a pretty standard  without extra performance issues? 

I'm just wondering on introducing new join type what the value we can bring up? 🤔

Reply:
<img width="748" alt="image" src="https://github.com/user-attachments/assets/76883c17-1bc8-43ef-aaaf-61a3dbaf8ecc">

Reply:
`it would be tempting to translate into a semi join` - they are 100% right, that exactly what I did 😄  That is interesting type of join, we def need to document its features

Reply:
We can not use just a `LeftOuter` because it will produce multiple rows per left row if there multiple matches in  the right input. (Note that this is what we do in master and this PR adds a test case that produce incorrect results on master) But we could do a distinct + LeftOuter to handle the duplicates but should be much slower than just doing a mark join.

Reply:
Thanks @eejbyfeldt, so it sounds to me like `LeftOuter + mark field + distinct`. I think we should describe the LeftMark join algorithm in test/example. 
For the usage is it planned to pick this algorithm automatically by the planner if some conditions met or the user currently calling it manually via API?

Reply:
It used/needed for decorrelating exists/in subqueries that are used in more complex boolean expressions like `OR`. This PR updates the optimizer rule decorrelate predicate subqueries to use mark join when it needed.